pkg/provision/cleanup.go

```diff
@@ -0,0 +1,250 @@
+				// delete the subvolume
+				log.Info().Msgf("deleting subvolume %s", subvol.Path)
+				if err := utils.SubvolumeRemove(ctx, filepath.Join(path, subvol.Path)); err != nil {
+					log.Err(err).Msgf("failed to delete subvol '%s'", subvol.Path)
+				}
+				if err := utils.QGroupDestroy(ctx, qgroup.ID, path); err != nil {
+					log.Err(err).Msgf("failed to delete qgroup: '%s'", qgroup.ID)
+				}
+				continue
+			}
+
+			// if the subvolume is a zdb and has 0 maxRfer, don't skip here. It might need to be deleted
+			zdbVolume := strings.HasPrefix(subvol.Path, "zdb") && qgroup.MaxRfer != 0
+
+			// we only handle volumes that are 256MiB or 10MiB
+			if qgroup.MaxRfer != 268435456 && qgroup.MaxRfer != 10485760 && !zdbVolume {
+				log.Info().Msgf("skipping volume '%s' is of size: %d", subvol.Path, qgroup.MaxRfer)
+				continue
+			}
+
+			// now, is this subvol in one of the toSave ?
+			if _, ok := toSave[filepath.Base(subvol.Path)]; ok {
+				log.Info().Msgf("skipping volume '%s' is used", subvol.Path)
+				continue
+			}
+
+			// delete the subvolume
+			log.Info().Msgf("deleting subvolume %s", subvol.Path)
+			if err := utils.SubvolumeRemove(ctx, filepath.Join(path, subvol.Path)); err != nil {
+				log.Err(err).Msgf("failed to delete subvol '%s'", subvol.Path)
+			}
+			if err := utils.QGroupDestroy(ctx, qgroup.ID, path); err != nil {
+				log.Err(err).Msgf("failed to delete qgroup: '%s'", qgroup.ID)
+			}
+		}
+	}
+
+	return nil
+}
+
+// checks running containers for subvolumes that might need to be saved because they are used
+// and subvolumes that might need to be deleted because they have no attached container anymore
+func checkContainers(msgBrokerCon string) (map[string]struct{}, map[string]struct{}, error) {
+	toSave := make(map[string]struct{})
+	toDelete := make(map[string]struct{})
+
+	for _, ns := range lines(must(execute("ctr", "namespace", "ls", "-q"))) {
+		log.Info().Msgf("Checking namespace %s", ns)
+		for _, cnt := range lines(must(execute("ctr", "-n", ns, "container", "ls", "-q"))) {
+			bytes := must(execute("ctr", "-n", ns, "container", "info", cnt))
+			var container container
+			if err := json.Unmarshal(bytes, &container); err != nil {
+				panic(err)
+			}
+
+			var namespaces []string
+			var err error
+
+			if ns == "zdb" {
+				zdbCl := initZdbConnection(container.ID)
+				defer zdbCl.Close()
+				if err := zdbCl.Connect(); err != nil {
+					log.Err(err).Msgf("failed to connect to 0-db: %s", container.ID)
+					continue
+				}
+
+				namespaces, err = zdbCl.Namespaces()
+				if err != nil {
+					log.Err(err).Msg("failed to retrieve zdb namespaces")
+					continue
+				}
+			}
+
+			toSave[filepath.Base(container.Spec.Root.Path)] = struct{}{}
+			for _, mnt := range container.Spec.Mounts {
+				if !strings.HasPrefix(mnt.Source, "/mnt/") {
+					continue
+				}
+
+				if len(ns) == 1 && namespaces[0] == "default" {
+					err := deleteZdbContainer(msgBrokerCon, pkg.ContainerID(container.ID))
```

Comment:
Do you still have to do that now that zdb decommission code already takes care of that ?

Reply:
Yeah, to clean up running containers that have no zdb volumes so we can shutdown the disks afterwards.

Reply:
But you already do that on the zdb code now, right? So why doing it here again?